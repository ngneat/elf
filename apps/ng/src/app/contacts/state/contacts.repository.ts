import { Injectable } from '@angular/core';
import { createState, Store } from '@ngneat/elf';

import {
  PaginationData,
  selectCurrentPage,
  selectCurrentPageEntities,
  setCurrentPage,
  setPage,
  updatePaginationData,
  withPagination,
} from '@ngneat/elf-pagination';
import { addEntities, withEntities } from '@ngneat/elf-entities';

interface Contact {
  id: number;
  email: string;
  name: string;
  address: string;
}

const { state, config } = createState(
  withPagination(),
  withEntities<Contact>()
);

const store = new Store({ state, config, name: 'contacts' });

@Injectable({ providedIn: 'root' })
export class ContactsRepository {
  activePage$ = store.pipe(selectCurrentPage());
  activePageContacts$ = store.pipe(selectCurrentPageEntities());

  setActivePage(id: Contact['id']) {
    store.update(setCurrentPage(id));
  }

  addContacts(page: number, contacts: PaginationData & { data: Contact[] }) {
    const { data, ...pagination } = contacts;

    store.update(
      addEntities(data),
      updatePaginationData(pagination),
      setPage(
        page,
        data.map((c) => c.id)
      )
    );
  }

  get store() {
    return store;
  }
}
