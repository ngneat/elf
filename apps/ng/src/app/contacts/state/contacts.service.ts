import { Injectable } from '@angular/core';
import { getData } from '../contacts-data';
import { timer } from 'rxjs';
import { map, tap } from 'rxjs/operators';
import { ContactsRepository } from './contacts.repository';
import { skipWhilePageExists } from '@ngneat/elf-pagination';

@Injectable({ providedIn: 'root' })
export class ContactsService {
  constructor(private repo: ContactsRepository) {}

  getContacts(page: number) {
    return timer(300).pipe(
      map(() => getData({ page })),
      tap((res) => this.repo.addContacts(page, res)),
      skipWhilePageExists(this.repo.store, page)
    );
  }
}
