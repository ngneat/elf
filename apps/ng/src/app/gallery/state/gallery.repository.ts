import { Injectable } from '@angular/core';
import { createStore } from '@ngneat/elf';
import {
  addEntities,
  selectAllEntities,
  selectEntities,
  UIEntitiesRef,
  unionEntities,
  updateEntities,
  withEntities,
  withUIEntities,
} from '@ngneat/elf-entities';

export interface GalleryItem {
  id: number;
  title: string;
  description: string;
  image: string;
}

export interface GalleryUIItem {
  id: number;
  open: boolean;
}

const store = createStore(
  { name: 'gallery' },
  withEntities<GalleryItem>(),
  withUIEntities<GalleryUIItem>()
);

@Injectable({ providedIn: 'root' })
export class GalleryRepository {
  items$ = this.store
    .combine({
      entities: store.pipe(selectAllEntities()),
      UIEntities: store.pipe(selectEntities({ ref: UIEntitiesRef })),
    })
    .pipe(unionEntities());

  addGalleryItems(galleryItems: GalleryItem[]) {
    const uiGalleryItems = galleryItems.map(({ id }) => ({ id, open: false }));
    store.update(
      addEntities(galleryItems),
      addEntities(uiGalleryItems, { ref: UIEntitiesRef })
    );
  }

  get store() {
    return store;
  }

  toggleItemOpen(galleryItemId: number) {
    this.store.update(
      updateEntities(galleryItemId, (item) => ({ ...item, open: !item.open }), {
        ref: UIEntitiesRef,
      })
    );
  }
}
