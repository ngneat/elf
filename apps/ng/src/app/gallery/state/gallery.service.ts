import { Injectable } from '@angular/core';
import { getData } from '../gallery-data';
import { timer } from 'rxjs';
import { map, tap } from 'rxjs/operators';
import { GalleryRepository } from './gallery.repository';

@Injectable({ providedIn: 'root' })
export class GalleryService {
  constructor(private repo: GalleryRepository) {}

  getGalleryItems() {
    return timer(300).pipe(
      map(() => getData()),
      tap((res) => this.repo.addGalleryItems(res))
    );
  }
}
