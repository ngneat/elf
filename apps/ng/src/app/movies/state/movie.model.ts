export type Movie = {
  id: number | string;
  title: string;
  genres: (string | number)[];
  actors: (string | number)[];
};

export interface FullMovie {
  id: number | string;
  title: string;
  genres: { id: number | string; name: string }[];
  actors: { id: number | string; name: string }[];
}
