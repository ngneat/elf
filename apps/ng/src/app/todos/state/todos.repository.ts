import { Injectable } from '@angular/core';
import { switchMap } from 'rxjs/operators';
import { write } from '../../store/mutations';
import { createState, select, Store, withProps } from '@ngneat/elf';
import {
  addEntities,
  selectAll,
  selectAllApply,
  updateEntities,
  withEntities,
} from '@ngneat/elf-entities';

interface Todo {
  id: number;
  title: string;
  completed: boolean;
}

export interface TodosProps {
  filter: 'ALL' | 'ACTIVE' | 'COMPLETED';
}

const { state, config } = createState(
  withEntities<Todo>(),
  withProps<TodosProps>({ filter: 'ALL' })
);

const store = new Store({ name: 'todos', state, config });

@Injectable({ providedIn: 'root' })
export class TodosRepository {
  todos$ = store.pipe(selectAll());
  filter$ = store.pipe(select((state) => state.filter));

  visibleTodos$ = this.filter$.pipe(
    switchMap((filter) => {
      return store.pipe(
        selectAllApply({
          filterEntity({ completed }) {
            if (filter === 'ALL') return true;
            return filter === 'COMPLETED' ? completed : !completed;
          },
        })
      );
    })
  );

  addTodo(title: Todo['title']) {
    store.update(addEntities({ id: Math.random(), title, completed: false }));
  }

  updateFilter(filter: TodosProps['filter']) {
    store.update(
      write((state) => {
        state.filter = filter;
      })
    );
  }

  updateCompleted(id: Todo['id']) {
    store.update(
      updateEntities(id, (entity) => ({
        ...entity,
        completed: !entity.completed,
      }))
    );
  }
}
