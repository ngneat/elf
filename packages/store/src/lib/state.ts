type UnionToIntersection<U> = (U extends any ? (k: U) => void : never) extends (
  k: infer I
) => void
  ? I
  : never;
type Merge<State extends any[], Key extends PropertyKey> = UnionToIntersection<
  State[number][Key]
>;

export type PropsFactory<Props, Config> = { props: Props; config: Config };
export type EmptyConfig = undefined;

export function createState<S extends PropsFactory<any, any>[]>(
  ...propsFactories: S
): { state: Merge<S, 'props'>; config: Merge<S, 'config'> } {
  const result = { config: {}, state: {} };

  for (const { config, props } of propsFactories) {
    Object.assign(result.config, config);
    Object.assign(result.state, props);
  }

  return result as { state: Merge<S, 'props'>; config: Merge<S, 'config'> };
}
