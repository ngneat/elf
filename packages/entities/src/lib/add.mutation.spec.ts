import { createStore } from '@ngneat/elf';
import {
  createEntitiesStore,
  createTodo,
  createUIEntityStore,
  createUITodo,
  toMatchSnapshot,
} from '@ngneat/elf-mocks';
import { addEntities, addEntitiesFifo } from './add.mutation';
import { UIEntitiesRef, withEntities, withUIEntities } from './entity.state';

describe('add', () => {
  let store: ReturnType<typeof createEntitiesStore>;

  beforeEach(() => {
    store = createEntitiesStore();
  });

  it('should add entity', () => {
    store.update(addEntities(createTodo(1)));
    toMatchSnapshot(expect, store, 'add one');
  });

  it('should add multiple entities', () => {
    store.update(addEntities([createTodo(1), createTodo(2)]));
    toMatchSnapshot(expect, store, 'add two');
  });

  it('should prepend entities', () => {
    store.update(addEntities([createTodo(1), createTodo(2)]));
    store.update(
      addEntities([createTodo(3), createTodo(4)], { prepend: true })
    );
    toMatchSnapshot(expect, store, 'prepend');
  });

  it('should work with ref', () => {
    const store = createUIEntityStore();
    store.update(
      addEntities([createUITodo(1), createUITodo(2)], { ref: UIEntitiesRef })
    );
    toMatchSnapshot(expect, store, 'ref');
  });

  it('should work a different idKey', () => {
    const store = createStore(
      { name: '' },
      withEntities<{ id: number }>(),
      withUIEntities<{ _id: string; name: string }, '_id'>({ idKey: '_id' })
    );
    store.update(
      addEntities({ _id: '1', name: 'foo' }, { ref: UIEntitiesRef })
    );
    store.update(addEntities({ id: 1 }));
    expect(store.getValue()).toMatchSnapshot();
  });

  describe('addEntitiesFifo', () => {
    it('should work', () => {
      const limit = 3;
      store.update(
        addEntitiesFifo([createTodo(1), createTodo(2), createTodo(3)], {
          limit,
        })
      );
      store.update(addEntitiesFifo([createTodo(4)], { limit }));

      expect(store.getValue()).toMatchSnapshot('should be 4 3 2');

      store.update(addEntitiesFifo([createTodo(5), createTodo(6)], { limit }));
      expect(store.getValue()).toMatchSnapshot('should be 6 5 4');

      store.update(
        addEntitiesFifo([createTodo(1), createTodo(2), createTodo(3)], {
          limit,
        })
      );
      expect(store.getValue()).toMatchSnapshot('should be 3 2 1');

      store.update(
        addEntitiesFifo(
          [createTodo(4), createTodo(5), createTodo(6), createTodo(7)],
          { limit }
        )
      );
      expect(store.getValue()).toMatchSnapshot('should be 7 6 5');

      store.update(addEntitiesFifo([], { limit }));
      expect(store.getValue()).toMatchSnapshot('should be 7 6 5 empty array');
    });
  });
});
