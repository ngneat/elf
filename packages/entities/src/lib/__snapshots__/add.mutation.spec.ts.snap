// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add addEntitiesFifo should work: should be 3 2 1 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "completed": false,
      "id": 1,
      "title": "todo 1",
    },
    "2": Object {
      "completed": false,
      "id": 2,
      "title": "todo 2",
    },
    "3": Object {
      "completed": false,
      "id": 3,
      "title": "todo 3",
    },
  },
  "ids": Array [
    1,
    2,
    3,
  ],
}
`;

exports[`add addEntitiesFifo should work: should be 4 3 2 1`] = `
Object {
  "entities": Object {
    "2": Object {
      "completed": false,
      "id": 2,
      "title": "todo 2",
    },
    "3": Object {
      "completed": false,
      "id": 3,
      "title": "todo 3",
    },
    "4": Object {
      "completed": false,
      "id": 4,
      "title": "todo 4",
    },
  },
  "ids": Array [
    2,
    3,
    4,
  ],
}
`;

exports[`add addEntitiesFifo should work: should be 6 5 4 1`] = `
Object {
  "entities": Object {
    "4": Object {
      "completed": false,
      "id": 4,
      "title": "todo 4",
    },
    "5": Object {
      "completed": false,
      "id": 5,
      "title": "todo 5",
    },
    "6": Object {
      "completed": false,
      "id": 6,
      "title": "todo 6",
    },
  },
  "ids": Array [
    4,
    5,
    6,
  ],
}
`;

exports[`add addEntitiesFifo should work: should be 7 6 5 1`] = `
Object {
  "entities": Object {
    "5": Object {
      "completed": false,
      "id": 5,
      "title": "todo 5",
    },
    "6": Object {
      "completed": false,
      "id": 6,
      "title": "todo 6",
    },
    "7": Object {
      "completed": false,
      "id": 7,
      "title": "todo 7",
    },
  },
  "ids": Array [
    5,
    6,
    7,
  ],
}
`;

exports[`add addEntitiesFifo should work: should be 7 6 5 empty array 1`] = `
Object {
  "entities": Object {
    "5": Object {
      "completed": false,
      "id": 5,
      "title": "todo 5",
    },
    "6": Object {
      "completed": false,
      "id": 6,
      "title": "todo 6",
    },
    "7": Object {
      "completed": false,
      "id": 7,
      "title": "todo 7",
    },
  },
  "ids": Array [
    5,
    6,
    7,
  ],
}
`;

exports[`add should add entity: add one 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "completed": false,
      "id": 1,
      "title": "todo 1",
    },
  },
  "ids": Array [
    1,
  ],
}
`;

exports[`add should add multiple entities: add two 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "completed": false,
      "id": 1,
      "title": "todo 1",
    },
    "2": Object {
      "completed": false,
      "id": 2,
      "title": "todo 2",
    },
  },
  "ids": Array [
    1,
    2,
  ],
}
`;

exports[`add should prepend entities: prepend 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "completed": false,
      "id": 1,
      "title": "todo 1",
    },
    "2": Object {
      "completed": false,
      "id": 2,
      "title": "todo 2",
    },
    "3": Object {
      "completed": false,
      "id": 3,
      "title": "todo 3",
    },
    "4": Object {
      "completed": false,
      "id": 4,
      "title": "todo 4",
    },
  },
  "ids": Array [
    3,
    4,
    1,
    2,
  ],
}
`;

exports[`add should work a different idKey 1`] = `
Object {
  "UIEntities": Object {
    "1": Object {
      "_id": "1",
      "name": "foo",
    },
  },
  "UIIds": Array [
    "1",
  ],
  "entities": Object {
    "1": Object {
      "id": 1,
    },
  },
  "ids": Array [
    1,
  ],
}
`;

exports[`add should work with ref: ref 1`] = `
Object {
  "UIEntities": Object {
    "1": Object {
      "id": 1,
      "open": false,
    },
    "2": Object {
      "id": 2,
      "open": false,
    },
  },
  "UIIds": Array [
    1,
    2,
  ],
}
`;
